op-batcher: add `channel.straddlesActivation` method (for Holocene) and wire it up

geoknee opened the pull request:
To trigger same flow as a channel timing out, when the channel confirmations straddle the Holocene activation timestamp

Closes https://github.com/ethereum-optimism/optimism/issues/12122

~⚠️ stacked on https://github.com/ethereum-optimism/optimism/pull/12390, don't merge yet.~

geoknee committed:
add channel.straddlesActivation method and wire it up

to trigger same flow as a channel timing out, when the channel confirmations straddle the Holocene activation timestamp

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -773,10 +777,22 @@ func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
-	l.state.TxConfirmed(id, l1block)
+	tctx, cancel := context.WithTimeout(l.shutdownCtx, l.Config.NetworkTimeout)
+	defer cancel()
+	header, err := l.L1Client.HeaderByNumber(tctx, receipt.BlockNumber)
+	if err != nil {
+		return err
+	}
```
There's an extra network call here to get the block timestamp, which we need to compare with the holocene time. 

It might be good to put some retry logic in here, `op-service` exposes some utilities for this. 

Currently if this network call fails, the channel timeout checks are not performed, even though in principle they still could be because the block number is still known.

sebastianst replied:
See my review, I think we should try indefinitely to get the header. It's important for future batcher operation that we correctly act on receipts.

vdamle commented on op-batcher/batcher/channel.go:
```diff
@@ -85,9 +85,9 @@ func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
-	// and then reset this state so it can try to build a new channel.
-	if c.isTimedOut() {
+	// If this channel timed out or if the confirmed tranactions straddle the Holocene activation time
+	// return true so that the the caller can reset the state abd build a new channel.
```
```suggestion
	// return true so that the the caller can reset the state and build a new channel.
```

vdamle commented on op-batcher/batcher/channel.go:
```diff
@@ -85,9 +85,9 @@ func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
-	if c.isTimedOut() {
+	// If this channel timed out or if the confirmed tranactions straddle the Holocene activation time
+	// return true so that the the caller can reset the state abd build a new channel.
+	if c.isTimedOut() || c.straddlesActivation(holoceneTime) {
```
Would we care whether the condition was triggered due to timeout or straddle? If so, maybe worth logging the correct reason? (This comment is actually for line 92)

geoknee replied:
Good call, fixed in [282740f](https://github.com/ethereum-optimism/optimism/pull/12743/commits/282740fb057ceaaed2855dcaf1e8b891afab22ed)

vdamle reviewed (commented):
lgtm overall, couple of minor nits.

geoknee committed:
typo

geoknee committed:
split up logging for channel timeout and channel straddle

geoknee committed:
add some retry logic and comments around receipt handling

geoknee committed:
typo

sebastianst commented on op-batcher/batcher/channel.go:
```diff
@@ -62,8 +62,8 @@ func (c *channel) TxFailed(id string) {
-// whether the channel timed out on chain.
-func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
+// whether the channel was invalidated.
+func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockRef, holoceneTime *uint64) bool {
```
I think it's worth adding the `rollup.Config` to the `channel` struct and then read the holocene activation time from there. The ctor already receives it and passes it to the `channelBuilder` ctor.

geoknee replied:
I figured we could actually just ready from the channelBuilder config: [a4e710d](https://github.com/ethereum-optimism/optimism/pull/12743/commits/a4e710da335a7ab9dd557a0af376ced1523e30fd).

sebastianst replied:
Yes we can do this as well, but I find this a bit abstraction-leaky... if the `channel` needs a rollup config, it should just store its own imo.

geoknee replied:
[130d704](https://github.com/ethereum-optimism/optimism/pull/12743/commits/130d70457a20fd28c1eb0017fb158fc350c227da)

sebastianst commented on op-batcher/batcher/channel.go:
```diff
@@ -85,11 +85,18 @@ func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
+
+	// If this channel had confirmed transactions straddling the Holocene activation time
+	// return true so that the the caller can reset the state and build a new channel.
+	if c.straddlesActivation(holoceneTime) {
```
Just thinking about the other PR in which we check for correct derivation progress. Should we have a function like `isInvalid()` around both checks. Otherwise we may have an edge case where a channel is invalid because it straddled the activation, but the derivation check wouldn't see it as invalid and assume that we should have progressed. Or is this a scenario that cannot happen?

geoknee replied:
The expected progress check will only throw an error if there is a _fully submitted_ channel which implies more progress should be made (current l1 advanced beyond max inclusion block, safe l2 head behind latest l2 block of that channel).

At the moment a channel becomes fully submitted, if it is invalid (for any reason, timing out or straddling holocene activation), it will be immediately and synchronously deleted by the `channelManager.handleChannelInvalidated()` method. So the expected progress check would never range over such an invalid channel.

sebastianst commented on op-batcher/batcher/channel.go:
```diff
@@ -62,8 +62,8 @@ func (c *channel) TxFailed(id string) {
-// whether the channel timed out on chain.
-func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) bool {
+// whether the channel was invalidated.
+func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockRef, holoceneTime *uint64) bool {
```
why did you change the type of `inclusionBlock` to a `BlockRef`?

geoknee replied:
See other thread https://github.com/ethereum-optimism/optimism/pull/12743#discussion_r1846425760

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -869,10 +879,24 @@ func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
+		ParentHash: header.ParentHash,
+		Time:       header.Time,
+	}
+	l.state.TxConfirmed(id, br)
```
OK I see where the nastiness comes from. Receipts, annoyingly, don't store the block time.

I think we shouldn't just do 3 tries here, but try to get the header in an infinite loop, and then not return an error. If the batcher L1 connection is broken, the whole batcher cannot operate. Otherwise, we may fail to record a receipt and never act on it, e.g. remove the channel, or see that the channel got invalidated. 

I remember that we do something similar in a different place of the batcher, but fail to find it right now.

Note that you could extract the common code that's also in `l1Tip` where a header is also retrieved and transformed into a `BlockRef` already.

geoknee replied:
[58b68e1](https://github.com/ethereum-optimism/optimism/pull/12743/commits/58b68e1308f0fa6a3c0b68ab8bf5f8ecb064ab31)

sebastianst reviewed (commented):
This approach works. Right now, we'd still submit frames of a straddled channel even if we're past the activation already and then only check when the confirmations arrives. This may be an acceptable trade off to optimize for simplicity vs frugality.

If it's not too complex, we could also fast-discard any open channels once the batcher crosses the L1 activation block in the main loop and sees that it still has channels open.

A test would also be good.

geoknee committed:
use existing rollupCfg to check holocene time

geoknee committed:
recordConfirmedTx tries forever to get L1Client.HeaderByNumber

does not return an error, converts header to blockref inline. Error removal bubbled up.

geoknee committed:
add TestChannelStraddlesActivation

geoknee commented:
> This approach works. Right now, we'd still submit frames of a straddled channel even if we're past the activation already and then only check when the confirmations arrives. This may be an acceptable trade off to optimize for simplicity vs frugality.
> 
> If it's not too complex, we could also fast-discard any open channels once the batcher crosses the L1 activation block in the main loop and sees that it still has channels open.
> 
> A test would also be good.

Test added here [42c3b91](https://github.com/ethereum-optimism/optimism/pull/12743/commits/42c3b91941d7fdafbfd095b2cfa68199390f4bbb). I think the current trade off is OK for detecting the straddling edge case -- as discussed it may not be worth the extra code to add optimization for discarding the a straddling channel before it is confirmed on chain.

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -901,8 +901,19 @@ func (l *BatchSubmitter) recordFailedTx(id txID, err error) {
+	var err error
+	for {
+		header, err = l.L1Client.HeaderByNumber(tctx, receipt.BlockNumber)
+		if err != nil {
+			break
+		}
```
Don't you need to break on `nil` error? We should also warn-log that we reattempt and put a pause or this can be a busy loop.
```suggestion
		if err == nil {
			break
		}
		l.log.Warn("Getting header for receipt failed, retrying.", "err", err)
		time.Sleep(2 * time.Second)
```
Wondering if there's also a generic function for this in the `retry` package.

I'm also wondering if there's an edge case where we receive a receipt, then an L1 reorg occurs and then we're getting the block time for a different block (but same block number), that might not include the receipt. We could fetch by block hash instead, but then the reorg case would fail here indefinitely...

Maybe the best we can do here is get by number, as it is now, but then check the block hash and in case it's different, we'd not confirm the tx but error-log and return early. However, the new block may also contain another receipt for the same transaction. We should check whether we'd get that new receipt as well by the txmgr logic. Then this code would be called again for the new receipt and we can safely just return in such a reorg case.

geoknee replied:
Good catch, fixed here [6d9adff](https://github.com/ethereum-optimism/optimism/pull/12743/commits/6d9adffd707be5b02eb838509e9ec906ddda85c0). 

The retry package does not allow us to retry indefinitely. We can either roll our own logic here or perhaps modify the retry package to have an `Indefinite` strategy.

From what I can tell about the txmgr: it will only return one receipt per transaction. So if that tx gets included, a receipt will be sent and then if there is an L1 reorg and the tx is included again (without being sent again by the txmgr) then there would not be a second receipt event in the batcher.

sebastianst commented:
> Test added here [42c3b91](https://github.com/ethereum-optimism/optimism/pull/12743/commits/42c3b91941d7fdafbfd095b2cfa68199390f4bbb). I think the current trade off is OK for detecting the straddling edge case -- as discussed it may not be worth the extra code to add optimization for discarding the a straddling channel before it is confirmed on chain.

OK agree, let's keep it simple but spendthrift for now! (Wanted to use an antonym even more fancy than frugal)

geoknee committed:
channel stores a rollupCfg ptr

geoknee committed:
fix headerByNumber retry logic

geoknee committed:
fix typo

geoknee commented:
We don't need this feature any more, since we rely on the existing "check expected progress" feature to handle this edge case. This was confirmed in Kurtosis testing here https://github.com/ethereum-optimism/optimism/issues/13192#issuecomment-2536542674.
